src/DynamoCore/Logging/DynamoAnalyticsClient.cs

```diff
@@ -167,102 +157,228 @@ public void Start()
         {
-            if (!ReportingAnalytics) return;
+            if (Analytics.DisableAnalytics) return;
 
-            var e = new ScreenViewEvent(viewName);
-            e.Track();
+            Task.Run(() =>
+            {
+                serviceInitialized.Wait();
+
+                lock (trackEventLockObj)
+                {
+                    if (!ReportingAnalytics) return;
+
+                    var e = new ScreenViewEvent(viewName);
+                    e.Track();
+                }
+            });
         }
 
         public void TrackException(Exception ex, bool isFatal)
         {
-            if (!ReportingAnalytics) return;
+            if (Analytics.DisableAnalytics) return;
+
+            Task.Run(() =>
+            {
+                serviceInitialized.Wait();
+
+                lock (trackEventLockObj)
+                {
+                    if (!ReportingAnalytics) return;
 
-            Service.TrackException(ex, isFatal);
+                    Service.TrackException(ex, isFatal);
+                }
+            });
         }
 
+        [Obsolete("Method will become private in Dynamo 4.0, please use CreateTaskTimedEvent")]
         public IDisposable CreateTimedEvent(Categories category, string variable, string description, int? value)
-        {
-            if (!ReportingAnalytics) return Disposable;
+        {            
+            serviceInitialized.Wait();
 
-            var e = new TimedEvent()
+            lock (trackEventLockObj)
             {
-                Category = category.ToString(),
-                VariableName = variable,
-                Description = description,
-                Value = value
-            };
-            //Timed event does not need startup tracking.
-            return e;
+                if (!ReportingAnalytics) return Disposable;
+
+                var e = new TimedEvent()
+                {
+                    Category = category.ToString(),
+                    VariableName = variable,
+                    Description = description,
+                    Value = value
+                };
+                //Timed event does not need startup tracking.
+                return e;
+            }
+        }
+
+        public Task<IDisposable> CreateTaskTimedEvent(Categories category, string variable, string description, int? value)
+        {
+            if (Analytics.DisableAnalytics) return Task.FromResult(Disposable);
+
+            return Task.Run(() => CreateTimedEvent(category, variable, description, value));
         }
 
+        [Obsolete("Property will become private in Dynamo 4.0, please use CreateTaskCommandEvent")]
         public IDisposable CreateCommandEvent(string name, string description, int? value)
         {
-            if (!ReportingAnalytics) return Disposable;
+            serviceInitialized.Wait();
+
+            lock (trackEventLockObj)
+            {
+                if (!ReportingAnalytics) return Disposable;
```

Comment:
I think we should try to minimize the amount of code executed within these locks.
Specifically I think we can move the `if (!ReportingAnalytics) return Disposable;` check above the lock (probably the same for all the locks)

Reply:
While ReportingAnalytics does only a null check on the service side right now that could change in the future, I think it doesn't really matter , it's safer to leave it as is.